[IMP] web, http, http_routing, *: new asset for test files

rdeodoo opened the pull request:
Coming with https://github.com/odoo/enterprise/pull/4281

This commit goal is to encapsulate every tour-test into a separate assets
bundle that would be called only during tests (command line) or by URL
(debug=tests). That way, a lot of .js files would not be loaded anymore
uselessly outside test mode and will speed up the page loads (especially in the
frontend as the backend do not reload the page anyway).

In order to do that, we needed to propagate the `debug` state from page to
page.
Otherwise, every page change during a test would simply lose the debug mode and
the test would stop as the test assets would not be loaded.
As we could not add the `&debug=tests` on every link (either hardcoded or
preprocess during the rendering), it has been decided to store it in session.

Technical summary:
1. `debug` state (currently only stored in URL) will be stored in session.
   Either when adding the `debug` param in URL (handled with _dispatch) or by
   starting Odoo with `test-enable` or `test-file` (handled by session init).
2. Once activated (and so set in session), debug mode will be remain activated
   even if not visible in URL (after a page navigation eg).
   To remove it,  set its value to nothing, `debug=`. That will exit debug mode
   (whatever mode it is: debug, assets, tests).
3. As tour-test files are now in a separate bundle, every layout (when needed)
   should `t-call="web.compiled_assets_tests"`.
   As it would be redundant and verbose, no `t-if` is needed on the t-call to
   load it only in test debug mode. That will be handled by
   `compiled_assets_tests` that will actually do the conditionnal t-call-assets
   to web.assets_tests, if tests debug mode is activated.
4. In addition to separating the tour-tests files in a separate bundle, we also
   moved those files to a specific folder under /static/tests/tours next to
   QUnit tests.
5. Also, tour files will be moved in a specific folder /static/src/js/tours for
   cleanness purpose (those files will still be kept in their 'normal' assets
   as needed outside test mode since it is tours).

rdeodoo commented:
I think I am good with this one for now.
@qsm-odoo Your input would be more that welcome, do you see any issue or unwanted behavior here?

Note that the commit about survey assets will be handled and backport by awa's team.
About the commits related to deprecated code and the x2many tour now ran without debug=assets I asked chm and wait for an answer.
Tests are now green (or were at least but can be considered as green), I just quickly created the enterprise PR that will also hopefully be green.

qsm-odoo commented on odoo/http.py:
```diff
@@ -64,6 +64,8 @@
 # To remove when corrected in Babel
 babel.core.LOCALE_ALIASES['nb'] = 'nb_NO'
 
+ALLOWED_DEBUG_MODE = [False, None, '', '1', 'assets', 'tests']
```
`ALLOWED_DEBUG_MODES` :)

rdeodoo replied:
:+1:

qsm-odoo commented on addons/survey/views/survey_templates.xml:
```diff
@@ -145,6 +145,7 @@
                     };
                 </script>
                 <t t-call-assets="web.assets_common"/>
+                <t t-call-assets="survey.minimal_assets"/>
```
Why? :)

rdeodoo replied:
Actually this is a bug since 12.1 but I asked for a fix that will be ported soon, it was done at https://github.com/odoo/odoo/pull/33218 so this one will be removed from this PR once ported and rebased.
But meanwhile this fix is needed on the PR for a green runbot otherwise since now tests are executed in debug mode tests (and thus debug mode), the error about duplicated files is not catch anymore, making the tour to fail.

qsm-odoo commented on addons/web/static/src/js/core/session.js:
```diff
@@ -301,9 +301,6 @@ var Session = core.Class.extend(mixins.EventDispatcherMixin, {
         options = _.clone(options || {});
         options.headers = _.extend({}, options.headers);
-        if (odoo.debug) {
-            options.headers["X-Debug-Mode"] = $.deparam($.param.querystring()).debug;
```
Just gone ? Not needed anymore ?

rdeodoo replied:
This was done to keep the debug state during RPC's, but since now the debug mode is stored in session and thus persistant, this is not needed anymore.

qsm-odoo commented on addons/web/views/webclient_templates.xml:
```diff
@@ -369,11 +378,16 @@
 
                 <t t-raw="head or ''"/>
             </head>
+            <div class="o_debug_mode text-center text-white" t-if="debug_mode" t-att-debug_mode="debug_mode">
```
I know JKE would not agree but I really prefer when the `t-if` is the first attribute... :)

rdeodoo replied:
Funny cuz I always change the t-if position multiple times since I can't decide which is the best, especially with that kind of class length size..
I'll change it since you are the first one to review, FIFO.

qsm-odoo commented on addons/web/views/webclient_templates.xml:
```diff
@@ -369,11 +378,16 @@
             </head>
+            <div class="o_debug_mode text-center text-white" t-if="debug_mode" t-att-debug_mode="debug_mode">
+                <t t-set="debug_mode_help" t-value="' (%s)' % debug_mode"/>
+                Debug mode is activated<t t-esc="debug_mode in ['assets', 'tests'] and debug_mode_help"/>. <a href="?debug=" class="text-dark">Exit debug mode.</a>
```
Why not put the condition in the definition of the variable :p ?

rdeodoo replied:
Good catch, I will change that

qsm-odoo commented on addons/web_tour/static/src/js/tour_manager.js:
```diff
@@ -142,7 +142,7 @@ return core.Class.extend(mixins.EventDispatcherMixin, ServicesMixin, {
 
-            var url = session.debug ? $.param.querystring(tour.url, {debug: session.debug}) : tour.url;
+            var url = tour.url;
             window.location.href = window.location.origin + url;
```
Lol :p Someone pays you to keep lines ? :p

rdeodoo replied:
Yeah, looking at the diff this looks stupid, did not pay attention while removing the ternary condition

qsm-odoo commented on addons/website/controllers/main.py:
```diff
@@ -350,9 +350,9 @@ def theme_customize(self, enable=None, disable=None, get_bundle=False):
             return {
-                'web.assets_common': request.env["ir.qweb"]._get_asset('web.assets_common', options=context),
-                'web.assets_frontend': request.env["ir.qweb"]._get_asset('web.assets_frontend', options=context),
-                'website.assets_editor': request.env["ir.qweb"]._get_asset('website.assets_editor', options=context),
```
Sure about that ?

rdeodoo replied:
This code was deprecated since v11, I will double check but I am sure there is a tour that test the theme customization right :hugs:?
My runbot is (almost) green :wink:

rdeodoo replied:
Actually this is not correct, it breaks theme customization.

But using that method is not good and should be changed, see https://github.com/odoo/odoo/commit/df4072b0b493000ccfa4e453d785bcea27659e5c.
I think I should remove this commit from this PR and it should be fixed on another PR, probably on lower version?

And since runbot is green with theme customization broken, I'd suggest to open a task to create a tour-tests about it.. This is a too big feature to be ignored..

qsm-odoo replied:
Why on a lower version ? There is no bug there right ?

rdeodoo replied:
I don't think there is a bug, what I said is just based on the deprecating commit, I just assumed things from its message.
Is there a plan to create tour(s) on theme customization? Or any reason not to?
Otherwise, I will remove that commit from this PR and create a task to do it in master and add a tour in the same time.

rdeodoo replied:
@qsm-odoo I moved deprecated commits to https://github.com/odoo/odoo/pull/33432

And about the task to create a tour, I think we should thank about which tours are needed and create a meta task to list all missing tours. Too much feature remains untested right now. Unsplash, Theme customize, Page property..
/cc @JKE-be @sbuhl

rdeodoo commented:
@qsm-odoo @JKE-be
I am done with this one. Rebased and runbot is green.

I followed both you advices, regarding JS debug_manager & http python.

This is the commit hierarchy of this PR:

> [IMP] web, http, http_routing: new asset for test files

This one is the core diff, small commit.

> [IMP] *: move test and tour files to new folder hierarchy and asset

This one is moving the files into different folder and assets.

> [IMP] test_new_api: don't run the test in debug assets

See commit message for this one, @JKE-be do you want me to squash it with another one? Which one?

> [IMP] project, website_slides(_forum): no need to forward debug in URL

Very small one, related to keep_query forwarding debug

> [IMP] web, web_tour, website, point_of_sale: adapt code to new debug

This one is the core diff of the framework level, @qsm-odoo could you review that one? Do you see another occurence I missed or something this could break?

> [IMP] web: rename file to prepare changes on next commit

Usefull commit to avoid next commit to have 1000 lines of diff.

> [IMP] mail,portal,web,web_tour,website: add debug manager in frontend

As discussed with @qsm-odoo, can you double check this one? It is quite small if you ignore the diff from the debug_manager.js that is almost just a move.
I moved the debug template into web module and included it in the /web/login

@JKE-be Could you, during the review, grep in your code the remaining occurence of `debug=`? Everything looks fine to me but just to be sure.
Also, about point_of_sale module, there is no indication about debug mode, I could add the web.debug_icon template next to the login name if you want. That's almost the only module layout without the debug icon (I even added it to mass_mailing unsubscribe template).
Lastly, I just noticed I should/could use `debug` and not `request.debug` in templates since debug is in qcontext (see [here](https://github.com/odoo-dev/odoo/blame/e8b69e5adce0447e1880754c684bebb7e963ce2c/odoo/addons/base/models/ir_ui_view.py#L1359)).

Thank you

rdeodoo commented:
@qsm-odoo @JKE-be I know you have a lot of stuff to do right now, but this one is a pain in the ass to maintain and source of 'bug' if people introduce new test meanwhile as they won't be in test_assets.

Once this one is merged, we will be able to send an email so people are aware of this and create test at the right place.
Also, we will be able to work on the logger.warning to detect misplaced files.

Thank you :)

Yenthe666 commented:
@rdeodoo I'm curious if I understand this correctly. Does this mean that all the JS files that contain tests - which are now also being compiled and rendered on a production db - would no longer be loaded?
So, this would downsize the size of the assets bundle and thus the loading speed of all website pages? 🎉

rdeodoo commented:
@Yenthe666 Hi, yes the goal is to encapsulate every .js test file in a separate bundle (and also move those to /tours folder so it will be even more clean).
That way, that asset won't be loaded anymore (except if you enable the debug=tests).

Yenthe666 commented:
Then this is without a doubt an awesome PR. Do you have any insights or stats about the difference in size of the assets bundle? Thanks for the fast feedback.

JKE-be commented:
~6000 of lines of Js removed from assets for everyone

+ thé ability to be in debug mode persistent... what was really painful in frontend ! Now it is stored!

rdeodoo commented:
Well it will of course depend of the number of app you have installed.
I checked the branch on runbot and this is the minified test bundle content: https://pastebin.com/Zj9bLJ08

odony commented:
@antonylesuisse you will want to have a look at this PR

odony commented on addons/web/models/ir_http.py:
```diff
@@ -4,14 +4,22 @@
+        # Store URL debug mode (might be empty) into session
+        if 'debug' in request.httprequest.args:
+            debug_mode = request.httprequest.args.get('debug')
+            request.session.debug = debug_mode if debug_mode in ALLOWED_DEBUG_MODES else True
```
don't alter the session unless you absolutely need it, ie. the debug mode has effectively changed

rdeodoo replied:
Good catch, the if condition is now a bit more verbose but done :)

antonylesuisse replied:
Why here ?

rdeodoo replied:
I moved it from `http_routing` to `web` as you can see on this [push force](https://github.com/odoo/odoo/compare/d3ace26a502a8515d3114f5bb69a2e05fba4d68d..d2c19fcf7b73374137d0eea0677ef1717a60bbc5) as this behavior is required even for non http_routing apps, else it would be inconsistent depending if the database as http_routing installed.

odony commented on odoo/http.py:
```diff
@@ -1049,6 +1052,9 @@ def _default_values(self):
         self.setdefault("context", {})
+        # Force tests debug mode when test mode enabled to load 'assets_tests'
+        test_mode = odoo.tools.config['test_enable'] or odoo.tools.config['test_file']
+        self.setdefault("debug", 'tests' if test_mode else False)
```
You would most likely need to adapt the logic of the [DisableCacheMiddleware](https://github.com/odoo/odoo/blob/e8b69e5adce0447e1880754c684bebb7e963ce2c/odoo/http.py#L1251-L1268) too

odony commented on odoo/http.py:
```diff
@@ -64,6 +64,8 @@
 # To remove when corrected in Babel
 babel.core.LOCALE_ALIASES['nb'] = 'nb_NO'
 
+ALLOWED_DEBUG_MODES = [False, None, '', '1', 'assets', 'tests']
```
It would be kind of weird to allow passing debug=1 to enable debug, but not debug=0 to disable it, don't you think? I imagine `'0'` could be aliased to `False` by `_dispatch()`, and `'1'` aliased to True?
Not sure about `None`, is it actually used with the current code paths? If unnecessary just omit it, so it's not tested every time.

rdeodoo replied:
Indeed, I used [str2bool](https://github.com/odoo-dev/odoo/blame/b3c6f95645d0c487000546c5ef3eb328e09da841/odoo/tools/misc.py#L440) and it will now manage such cases logically.
That allowed also to clean `ALLOWED_DEBUG_MODES` to only [False, True, 'assets', 'tests']

antonylesuisse commented:
I'm ok with defining assets_test, i'm against the rest, especially moving debug into session you should be able to have 2 tabs open one in debug one without.

JKE-be commented:
@antonylesuisse 

Are you in GR ? 

Let me know in which case you need to have one tab in debug and the other not?
Let me know when you are in the frontend e.g. on /shop/checkout with debug_asset=1 and want to submit a form, you don't want to keep the asset... ? 

Or we can do like in backend one case by one, in this case I want to propagate the debug... It is really not a good practice imo. It is exactly like the kw in controller., when we add it, it is too late, it is because we add a bug before!

It is strange to don't have any pov from you during all the dev, to have the idea validate by chs et odoo and that after you want to remove it.

Moving debug in session is imo a real improvement and cleaning of the code.
And make the life simple for everybody.

JKE-be commented:
Seems like this task was a stupid idea ... my bad!

antonylesuisse commented on addons/point_of_sale/views/pos_templates.xml:
```diff
@@ -31,12 +31,14 @@
         <t t-call-assets="web.assets_common" t-css="false"/>
         <t t-call-assets="point_of_sale.pos_assets_backend" t-css="false"/>
         <t t-call-assets="point_of_sale.assets"/>
+        <t t-call="web.compiled_assets_tests"/>
```
optional_assets_tests ?
test_assets_tests

rdeodoo replied:
@qsm-odoo Maybe you can give your input here about the template name? I think that's the nomenclature used by editor and studio for wrapper template?

rdeodoo replied:
Every wrapper template currently written with `_compiled` prefix, see https://github.com/odoo-dev/odoo/blame/b3c6f95645d0c487000546c5ef3eb328e09da841/addons/web_editor/views/editor.xml#L9
https://github.com/odoo-dev/odoo/blame/b3c6f95645d0c487000546c5ef3eb328e09da841/addons/website/views/website_templates.xml#L89
https://github.com/odoo-dev/enterprise/blame/254c2062b6c493531b4d17166908139fbf847848/web_studio/views/assets.xml#L103

I think it should remain like that for clearness purpose

qsm-odoo replied:
Yes, probably not a perfect name but at least it is consistent with the rest.

JKE-be replied:
final name: conditional_assets_test

+rename 3 other with the same convention

rdeodoo replied:
I already renamed conditional_assets_test but I won't rename the other since they are not conditional but wrapper (no explicit group or if condition on it).
I don't like that and I prefered _compiled but it is done :)

antonylesuisse commented on addons/web/controllers/main.py:
```diff
@@ -1555,8 +1549,7 @@ def get_app_url(self, req, app):
         sakey = Session().save_session_action(action)
-        debug = '?debug' if req.debug else ''
-        return werkzeug.utils.redirect('/web{0}#sa={1}'.format(debug, sakey))
+        return werkzeug.utils.redirect('/web#sa={1}'.format(sakey))
```
did you test ?

rdeodoo replied:
Hm, obviously not since `{1}` does not exist, my bad, I will change that.
Anyway, I just had a quick look at this feature and it seems to not work anymore? Last working version I found is 11.0, later version will keep redirecting on /web and not module form view.

rdeodoo replied:
Behavior fixed in saas-11.3 with https://github.com/odoo/odoo/commit/e9350993ca0b81df17308766d7e26d458bc39520

antonylesuisse commented on addons/web/views/webclient_templates.xml:
```diff
@@ -446,8 +458,9 @@
                 <div t-attf-class="clearfix oe_login_buttons text-center mb-1 {{'pt-2' if form_small else 'pt-3'}}">
                     <button type="submit" class="btn btn-primary btn-block">Log in</button>
-                    <t t-if="debug">
+                    <t t-if="request.debug">
```
request.session

rdeodoo commented on addons/web_settings_dashboard/static/src/xml/dashboard.xml:
```diff
@@ -123,15 +123,13 @@
-            <br t-if="debug != false"/>
-            <a t-if="(debug) and !(demo_active)" class="oe_activate_debug_mode float-right o_web_settings_dashboard_force_demo" href="#">Load demo data</a>
+            <div class="text-right">
+                <a t-if="debug.indexOf('1') == -1" class="d-block" href="?debug=1" >Activate the developer mode</a>
```
It is not needed here as debug is actually coming from request.session.debug (which is always a string), it is not the odoo.debug js variable, even if this is static, see [here](https://github.com/odoo/odoo/blame/079fcd6232e0c0338c353e6281dc5a930aa91e3a/addons/web_settings_dashboard/controllers/main.py#L76)

rdeodoo commented:
All feedbacks got processed, included `DisableCacheMiddleware`.
The branch has been rebased and is green.
Same for enterprise.

JKE-be commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

JKE-be commented:
@robodoo r-

JKE-be commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

JKE-be commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
Linked pull request(s) odoo/enterprise#4281 not ready. Linked PRs are not staged until all of them are ready.

JKE-be commented:
@robodoo retry

robodoo commented:
I'm sorry, @JKE-be. Retry makes no sense when the PR is not in error.

JKE-be commented:
@robodo retry

JKE-be commented:
@robodoo r+

JKE-be commented:
@robodoo retry

robodoo commented:
I'm sorry, @JKE-be. Retry makes no sense when the PR is not in error.

qsm-odoo commented:
@robodoo delegate+

rdeodoo commented:
@robodoo r+ rebase-merge

robodoo commented:
I'm sorry, @rdeodoo. You can't method=rebase-merge.

robodoo commented:
Unable to stage PR ({"message":"Merge conflict","documentation_url":"https://developer.github.com/v3/repos/merging/#perform-a-merge"})

JKE-be commented:
@robodoo r+ rebase-merge retry p=1

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
I'm sorry, @JKE-be. Retry makes no sense when the PR is not in error.

robodoo commented:
Merged, thanks!
